packages/arb-rpc-node/batcher/batcher.go

```diff
@@ -237,46 +215,95 @@ func newBatcher(
 	return server
 }
 
-func (m *Batcher) sendBatch(ctx context.Context, inbox l2TxSender) {
+func (m *Batcher) handleNextTx() bool {
+	tx, accountIndex, cont := popRandomTx(m.pendingBatch, m.queuedTxes)
+	if tx != nil {
+		err := m.pendingBatch.addIncludedTx(tx)
+		m.queuedTxes.maybeRemoveAccountAtIndex(accountIndex)
+		if err != nil {
+			logger.Error().Err(err).Msg("Aggregator ignored invalid tx")
+		}
+	}
+	return cont
+}
+
+func (m *Batcher) maybeSubmitBatch(ctx context.Context, maxBatchTime time.Duration, lastBatch time.Time, globalInbox l2TxSender, moreTxesWaiting bool) (bool, error) {
 	txes := m.pendingBatch.getAppliedTxes()
-	if len(txes) == 0 {
-		return
+	full := m.pendingBatch.isFull()
+	m.Unlock()
+
+	if !full && !(len(txes) > 0 && !moreTxesWaiting && time.Since(lastBatch) > maxBatchTime) {
+		return false, nil
 	}
+	lastBatch = time.Now()
 	batchTxes := make([]message.AbstractL2Message, 0, len(txes))
 	for _, tx := range txes {
 		batchTxes = append(batchTxes, message.NewCompressedECDSAFromEth(tx))
 	}
 	batchTx, err := message.NewTransactionBatchFromMessages(batchTxes)
 	if err != nil {
-		logger.Fatal().Err(err).Msg("transaction aggregator failed")
+		return false, errors.Wrap(err, "invalid transaction in batch")
 	}
-	for {
-		logger.Info().Int("txcount", len(batchTxes)).Msg("Submitting batch")
-		txHash, err := inbox.SendL2MessageFromOrigin(
-			ctx,
-			message.NewSafeL2Message(batchTx).AsData(),
-		)
 
-		if err != nil {
-			time.Sleep(5 * time.Second)
-			logger.Error().Err(err).Msg("error calling SendL2MessageFromOrigin, retrying")
-			continue
-		}
+	logger.Info().Int("txcount", len(txes)).Msg("Submitting batch")
+	batchData := message.NewSafeL2Message(batchTx).AsData()
+	txHash, err := globalInbox.SendL2MessageFromOrigin(ctx, batchData)
+	if err != nil {
+		return false, errors.Wrap(err, "error calling SendL2MessageFromOrigin")
```

Comment:
Just double checking that the removal of retry logic here is intentional (maybe it was added somewhere else and I missed it?)

Reply:
So there was retry logic both here and in the code above this code. I decided that it would be nicer to flatten out the code and have a single codepath for retries in order to reduce the level of nesting of the code